clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -609,7 +607,18 @@ private void handleProduceResponse(ClientResponse response, Map<TopicPartition,
                                 .collect(Collectors.toList()),
                             p.errorMessage(),
                             p.currentLeader());
-                    ProducerBatch batch = batches.get(tp);
+                    // Version 13 drop topic name and add support to topic id.
+                    // We need to find batch based on topic id and partition index only as
+                    // topic name in the response might be empty.
+                    TopicIdPartition tpId = new TopicIdPartition(r.topicId(), p.index(), r.name());
+                    ProducerBatch batch = batches.entrySet().stream()
```

Comment:
This changes a map lookup to an iteration. Could we do some produce perf test (with multiple topic/partitions) to verify there is no performance degradation?

Reply:
An alternative is for `handleProduceResponse()` to take a Map<TopicPartition, ProducerBatch> and a Map<UUID, String>. If the response has non-zero topicId, we look up the second map to find the topic name and then use the first map to find the batch. Otherwise, we look up the first map using the topic name.